File: demos_code_J2_conception_objet/08_executable_architecture_rules/DemoArchitectureRules.java
import java.util.*;

public class DemoArchitectureRules {
    record Dependency(String from, String to) { }

    static final Set<String> FORBIDDEN_FOR_DOMAIN = Set.of("infrastructure", "web", "stripe");

    static void assertRules(List<Dependency> deps) {
        for (Dependency d : deps) {
            if (d.from().equals("domain") && FORBIDDEN_FOR_DOMAIN.contains(d.to())) {
                throw new AssertionError("Règle violée : domain ne doit pas dépendre de " + d.to());
            }
        }
        assertAcyclic(deps);
    }

    static void assertAcyclic(List<Dependency> deps) {
        Map<String, List<String>> graph = new HashMap<>();
        for (Dependency d : deps) graph.computeIfAbsent(d.from(), k -> new ArrayList<>()).add(d.to());
        Set<String> visiting = new HashSet<>(), visited = new HashSet<>();
        for (String node : graph.keySet()) dfs(node, graph, visiting, visited);
    }

    static void dfs(String node, Map<String, List<String>> g, Set<String> visiting, Set<String> visited) {
        if (visiting.contains(node)) throw new AssertionError("Cycle détecté autour de " + node);
        if (!visited.add(node)) return;
        visiting.add(node);
        for (String next : g.getOrDefault(node, List.of())) dfs(next, g, visiting, visited);
        visiting.remove(node);
    }

    public static void main(String[] args) {
        List<Dependency> valid = List.of(
                new Dependency("web", "application"),
                new Dependency("application", "domain"),
                new Dependency("infrastructure", "application")
        );
        assertRules(valid);
        System.out.println("Architecture valide ✓");

        List<Dependency> invalid = List.of(
                new Dependency("domain", "stripe"),
                new Dependency("application", "domain")
        );
        try {
            assertRules(invalid);
        } catch (AssertionError e) {
            System.out.println("Architecture refusée ✗ : " + e.getMessage());
        }

        List<Dependency> cyclic = List.of(
                new Dependency("orders", "billing"),
                new Dependency("billing", "orders")
        );
        try {
            assertRules(cyclic);
        } catch (AssertionError e) {
            System.out.println("Cycle refusé ✗ : " + e.getMessage());
        }
    }
}

File: Demo en Python/demos_code_J2_conception_objet_python/08_executable_architecture_rules/demo_architecture_rules.py
from dataclasses import dataclass
from collections import defaultdict


@dataclass(frozen=True)
class Dependency:
    source: str
    target: str


FORBIDDEN_FOR_DOMAIN = {"infrastructure", "web", "stripe"}


def assert_rules(dependencies: list[Dependency]) -> None:
    for dependency in dependencies:
        if dependency.source == "domain" and dependency.target in FORBIDDEN_FOR_DOMAIN:
            raise AssertionError(
                f"Règle violée : domain ne doit pas dépendre de {dependency.target}"
            )
    assert_acyclic(dependencies)


def assert_acyclic(dependencies: list[Dependency]) -> None:
    graph: dict[str, list[str]] = defaultdict(list)
    for dependency in dependencies:
        graph[dependency.source].append(dependency.target)

    visiting: set[str] = set()
    visited: set[str] = set()

    def dfs(node: str) -> None:
        if node in visiting:
            raise AssertionError(f"Cycle détecté autour de {node}")
        if node in visited:
            return

        visiting.add(node)
        for next_node in graph.get(node, []):
            dfs(next_node)
        visiting.remove(node)
        visited.add(node)

    for node in list(graph):
        dfs(node)


if __name__ == "__main__":
    valid = [
        Dependency("web", "application"),
        Dependency("application", "domain"),
        Dependency("infrastructure", "application"),
    ]
    assert_rules(valid)
    print("Architecture valide ✓")

    invalid = [
        Dependency("domain", "stripe"),
        Dependency("application", "domain"),
    ]
    try:
        assert_rules(invalid)
    except AssertionError as error:
        print("Architecture refusée ✗ : " + str(error))

    cyclic = [
        Dependency("orders", "billing"),
        Dependency("billing", "orders"),
    ]
    try:
        assert_rules(cyclic)
    except AssertionError as error:
        print("Cycle refusé ✗ : " + str(error))
